test_client/src/main/java/io/reactivesocket/tck/ParseMarble.java

```diff
@@ -35,117 +39,101 @@ public ParseMarble(String marble, Subscriber<? super Payload> s) {
-                default:
-                    if (argMap != null) {
-                        // this is hacky, but we only expect one key and one value
-                        Map<String, String> tempMap = argMap.get(c + "");
-                        if (tempMap == null) {
-                            s.onNext(new PayloadImpl(c + "", c + ""));
-                            break;
-                        }
-                        List<String> key = new ArrayList<>(tempMap.keySet());
-                        List<String> value = new ArrayList<>(tempMap.values());
-                        s.onNext(new PayloadImpl(key.get(0), value.get(0)));
-                    }
-                    else s.onNext(new PayloadImpl(c + "", c + "")); // if value not mapped, just send as data and meta
-                    numSent++;
-                    break;
-            }
-            iter.next();
-        }
+    // adds stuff to the end of marble
+    public synchronized void add(String m) {
+        System.out.println("adding " + m);
+        this.marble += m;
+        parseLatch.countDown();
     }
 
-    /*private void parseMarble(String marble, Subscriber<? super Payload> s) {
-        String buffer = "";
-        boolean grouped = false;
+    public synchronized void request(long n) {
+        System.out.println("requested" + n);
+        numRequested += n;
+        if (marble.length() > marbleIndex) sendLatch.countDown();
+    }
 
+    // this parses the actual marble diagram and acts out the behavior
+    // should be called upon triggering a handler
+    public void parse() {
         try {
-            char[] commands = marble.toCharArray();
-            Map<String, Map<String, String>> argMap = null;
-            if (marble.contains("&&")) {
-                String[] temp = marble.split("&&");
-                commands = temp[0].toCharArray();
-                ObjectMapper mapper = new ObjectMapper();
-                argMap = mapper.readValue(temp[1], new TypeReference<Map<String, Map<String, String>>>(){});
-            }
-            for (char c : commands) {
+            // if cancel has been called, don't do anything
+            if (cancelled) return;
+            String buffer = "";
+            boolean grouped = false;
+            while (true) {
+                if (marbleIndex >= marble.length()) {
+                    if (parseLatch.getCount() == 0) parseLatch = new CountDownLatch(1);
+                    parseLatch.await();
+                    parseLatch = new CountDownLatch(1);
+                }
+                char c = marble.charAt(marbleIndex);
                 switch (c) {
                     case '-':
                         if (grouped) buffer += c;
-                        else Thread.sleep(10);
+                        else try {
+                            Thread.sleep(10);
+                        } catch (Exception e) {
+                            System.out.println("Interrupted");
+                        }
```

Comment:
I actually will probably get rid of this... I was getting a bit mixed up between Rx and reactive sockets. Rx apparently has virtual time units, which is fine for testing timing, but with reactive sockets any tests that use timing will be nondeterministic